storage/src/main/java/com/tencent/rss/storage/handler/impl/ComposedClientReadHandler.java

```diff
@@ -47,52 +56,102 @@ public ComposedClientReadHandler(ClientReadHandler... handlers) {
     if (size > 2) {
       this.coldDataReadHandler = handlers[2];
     }
+    if (size > 3) {
+      this.frozenDataReadHandler = handlers[3];
+    }
+  }
+
+  public ComposedClientReadHandler(Callable<ClientReadHandler>... creators) {
+    int size = creators.length;
+    if (size > 0) {
+      this.hotHandlerCreator = creators[0];
+    }
+    if (size > 1) {
+      this.warmHandlerCreator = creators[1];
+    }
+    if (size > 2) {
+      this.coldHandlerCreator = creators[2];
+    }
+    if (size > 3) {
+      this.frozenHandlerCreator = creators[3];
+    }
   }
 
   @Override
   public ShuffleDataResult readShuffleData() {
     ShuffleDataResult shuffleDataResult = null;
-    switch (currentHandler) {
-      case HOT:
-        try {
+    try {
+      switch (currentHandler) {
+        case HOT:
+          if (hotDataReadHandler == null) {
+            hotDataReadHandler = createReadHandlerIfNotExist(hotHandlerCreator);
+            hotDataReadHandler = hotHandlerCreator.call();
```

Comment:
it should be called in createReadHandlerIfNotExist already

Reply:
sorry, i will remove it